storage/innobase/row/row0ins.cc

```diff
@@ -1568,6 +1578,69 @@ class ib_dec_in_dtor {
 	ulint&		counter;
 };
 
+/*********************************************************************//**
+Reads sys_trx_end field from clustered index row.
+@return trx_id_t */
+static
+trx_id_t
+row_ins_get_sys_trx_end(
+/*===================================*/
+			const rec_t *rec,	/*!< in: clustered row */
+			ulint *offsets,		/*!< in: offsets */
+			dict_index_t *index)	/*!< in: clustered index */
+{
+	ut_a(dict_index_is_clust(index));
+
+	ulint len;
+	ulint nfield = dict_col_get_clust_pos(
+		&index->table->cols[index->table->vers_row_end], index);
+	const byte *field = rec_get_nth_field(rec, offsets, nfield, &len);
+	ut_a(len == 8);
+	return(mach_read_from_8(field));
+}
+
+/*********************************************************************//**
+Performs search at clustered index and returns sys_trx_end if row was found.
+@return DB_SUCCESS, DB_NO_REFERENCED_ROW */
+static
+dberr_t
+row_ins_search_sys_trx_end(
+/*=======================*/
+			dict_index_t *index,	/*!< in: index of record */
+			const rec_t *rec,	/*!< in: record */
+			trx_id_t *end_trx_id)	/*!< out: end_trx_id */
+{
```

Comment:
The comment and the function name should convey that `index` and `rec` are in a secondary index. Also, add `ut_ad(!index->is_clust())` at the start of the function.

Reply:
https://github.com/tempesta-tech/mariadb/issues/306